biojava-structure-gui/src/main/java/org/biojava/nbio/structure/align/gui/jmol/JmolPanel.java

```diff
@@ -114,11 +127,28 @@ public void executeCmd(String rasmolScript) {
 		viewer.evalString(rasmolScript);
 	}
 
-	public void setStructure(Structure s)
+	public void setStructure(final Structure s)
 	{
 		this.structure = s;
-		String serialized = s.toMMCIF();
-		viewer.openStringInline(serialized);
+		try (
+				PipedOutputStream out = new PipedOutputStream();
+				// Viewer requires a BufferedInputStream for reflection
+				InputStream in = new BufferedInputStream(new PipedInputStream(out));
+				) {
+			new Thread((Runnable)() -> {
+				try {
+					MmtfActions.writeToOutputStream(s,out);
+				} catch (Exception e) {
+					logger.error("Error generating MMTF output for {}",
+							s.getStructureIdentifier()==null ? s.getStructureIdentifier().getIdentifier() : s.getName(), e);
+				}
+			}).start();
+			viewer.openReader(null, in);
+		} catch (IOException e) {
```

Comment:
The problem is that the interesting exception is the one from writeToOutputStream, and that has to run in its own thread if I want to use pipes like this, so it can't throw an exception.

The other IOExceptions would indicate a broken pipe to Jmol, which would very likely be from program termination anyways. Trying to throw an IOException here requires adding the exception to maybe a dozen other functions that call this, some of which don't make much sense.

I guess I could write the MMTF to a byte[] and then pass it to jmol (avoiding creating a new thread).

I guess ultimately I caught the exceptions here because they seemed very unlikely to result from user input (assuming the MMTF encoder doesn't have bugs).